### PR TITLE
NTFS, enhance LZNT1 decompression with  more boundary checks and bit …

### PR DESCRIPTION
**Pull request by Tony-Nuix**
Modify code according to https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-xca/5655f4a3-6ba4-489b-959f-e1f407c52f15

### CODE REVIEW

File: src/main/java/org/jnode/fs/ntfs/datarun/CompressedDataRun.java [hunk 1]
```diff
@@ -220,43 +248,96 @@ private static int uncompressBlock(final OffsetByteArray compressed,
...
+                    // In other words, it is not required that (U – displacement + length < U).
+                    // Therefore, when processing a compressed word, data MUST be copied from the start of the uncompressed target region to the end—that is,
+                    // the byte at (U – displacement) MUST be copied first, then (U – displacement + 1), and so on,
+                    // because the compressed word might refer to data that will be written during decompression.
+
+                    // Also, the Linux code is https://github.com/torvalds/linux/blob/master/fs/ntfs3/lznt.c#L275-L277 as a reference.
+
+                    // DO NOT use arrayCopy in any case.
+                    for (int k = 0; k < blen; k++) {
+                        uncompressed.put(pos, uncompressed.get(pos - boff));
```
Comment (Tony-Nuix) [inline src/main/java/org/jnode/fs/ntfs/datarun/CompressedDataRun.java:350]: @tmyroadctfig 
Please ignore the other PR (which is for investigation only) 
This line is the main change. 
The uncompressed.copyFrom somehow has some problem. and it's better (and easier to be implemented) to just always copy the bytes one by one.

File: src/main/java/org/jnode/fs/ntfs/datarun/CompressedDataRun.java [hunk 2]
```diff
@@ -220,43 +248,96 @@ private static int uncompressBlock(final OffsetByteArray compressed,
...
             // Uncompressed chunks store length as 0xFFF always.
             if ((len + 1) != BLOCK_SIZE) {
                 log.debug("ntfs_uncompblock: len: " + len + " instead of 0xfff");
             }
 
-            // Copies the entire compression block as-is, need to skip the compression flag,
-            // no idea why they even stored it given that it isn't used.
-            // Darwin's version I was referring to doesn't skip this, which seems be a bug.
-            uncompressed.copyFrom(compressed, cpos, 0, len + 1);
-            uncompressed.zero(len + 1, BLOCK_SIZE - 1 - len);
```
Comment (Tony-Nuix) [inline src/main/java/org/jnode/fs/ntfs/datarun/CompressedDataRun.java:233]: Is this necessary? Needs more tests to confirm whether it still needs to be filled with zero or not.
hmmm, maybe still yes.. then I need to rollback it here.

File: src/main/java/org/jnode/fs/ntfs/datarun/CompressedDataRun.java [hunk 3]
```diff
@@ -220,43 +248,96 @@ private static int uncompressBlock(final OffsetByteArray compressed,
...
             return len + 3;
         }
 
-        while (cpos < len + 3 && pos < BLOCK_SIZE) {
+        int rightmostInUncompressed = Math.min(uncompressed.array.length, BLOCK_SIZE);
+
+        // Now this chunk contains compressed data, decompress it.
+        while (cpos < len + 3 && pos < rightmostInUncompressed) {
+            // https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-xca/1fd21d29-f42f-4fc4-b677-9de7cc386be8
+            // If a chunk is compressed, its chunk header is immediately followed by the first byte of a Flag_group nonterminal.
```
Comment (Tony-Nuix) [inline src/main/java/org/jnode/fs/ntfs/datarun/CompressedDataRun.java:283]: All the comments are copied from Microsoft official website. I feel it is a good idea to make the code crystal clear, though the comments make the code messy.

File: src/test/java/org/jnode/fs/FileSystemTestUtils.java [hunk 4]
```diff
@@ -101,4 +84,46 @@ public static byte[] intsToByteArray(String signedInts) {
...
+
+    /**
+     * Reads a file and write content into a byte array.
+     *
+     * @param filePath the file path.
+     * @return the byte array
+     * @throws IOException if any error occurs.
+     */
+    @Nonnull
+    public static byte[] readFileToByteArray(String filePath) throws IOException {
```
Comment (tmyroadctfig) [inline src/test/java/org/jnode/fs/FileSystemTestUtils.java:97]: Couldn't this be replaced with `Files.readAllBytes()`?

File: src/test/java/org/jnode/fs/FileSystemTestUtils.java [hunk 5]
```diff
@@ -101,4 +84,46 @@ public static byte[] intsToByteArray(String signedInts) {
...
         return bytes;
     }
+
+    /**
+     * Converts a string of signed ints to a byte array.
+     *
+     * @param ints the int array.
+     * @return the byte array.
+     */
+    public static byte[] intArrayToByteArray(int[] ints) {
```
Comment (tmyroadctfig) [inline src/test/java/org/jnode/fs/FileSystemTestUtils.java]: I think we should just save the raw data to a file instead of keeping the `int[]` data in `CompressedTestData`

File: src/test/java/org/jnode/fs/ntfs/NTFSCompressedDataRunTriage4924Test.java [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+        // Assert
+        byte[] expectedByteArray = {
+                40, 126, 9, 0, 6, 114, 32, -79, 0, 112, 25, -115, 28,
+                0, 0, 1, 37, 22, 17, 8, 111, 56, 19, 0, 6, 114, -15,
+                31, 0, 112, 111, 49, 1, 0, 10, -94, 37, 23, 17, 8,
+                111, 53, 19, 0, 6, 17, 8, 111, 56, 19, 0, 6, 111,
+                -95, 0, 0, 10, 126, 112, 0, 0, 10, 111, -26, 2, 0,
+                10, -94, 37, 24, 9, 123, 4, 8, 0, 4, 111, 53, 19,
+                0, 6, 9, 123, 4, 8, 0, 4, 111, 56, 19, 0, 6, 111,
+                -95, 0, 0, 10, 126, 112, 0
```
Comment (tmyroadctfig) [inline src/test/java/org/jnode/fs/ntfs/NTFSCompressedDataRunTriage4924Test.java]: Maybe we can write the expected data into an "expected" file. E.g. `"org/jnode/fs/ntfs/offset-before-start.compressed_expected.bin"`

File: src/test/resources/org/jnode/fs/ntfs/magnet-forensics/block1.uncompressed.bin
Comment (tmyroadctfig) [inline src/test/resources/org/jnode/fs/ntfs/magnet-forensics/block1.uncompressed.bin:1]: Is the name of this mixed up with the compressed one?

Comment (Tony-Nuix) [reply src/test/resources/org/jnode/fs/ntfs/magnet-forensics/block1.uncompressed.bin:1]: It's copied from https://github.com/MagnetForensics/rust-lzxpress/tree/main/tests directly. But I can't pass the test even if I run their c code directly. We may just ignore this test data for now.

File: src/test/java/org/jnode/fs/FileSystemTestUtils.java (hunk 5 above)
Comment (Tony-Nuix) [reply src/test/java/org/jnode/fs/FileSystemTestUtils.java]: The test data is coped from https://github.com/libyal/libfsntfs/blob/82181db7c9f272f98257cf3576243d9ccbbe8823/tests/fsntfs_test_compression.c#L39 directly. I feel it is easier to compare them if I keep the format unchanged. (E.g. if they update the data later, we can easily apply to our test code here.)

File: src/test/java/org/jnode/fs/FileSystemTestUtils.java (hunk 4 above)
Comment (tmyroadctfig) [reply src/test/java/org/jnode/fs/FileSystemTestUtils.java:97]: You still need to tidy away the commented code below

File: src/test/java/org/jnode/fs/FileSystemTestUtils.java (hunk 5 above)
Comment (tmyroadctfig) [reply src/test/java/org/jnode/fs/FileSystemTestUtils.java]: But the test should never really change because the algorithm is not changing

File: src/main/java/org/jnode/fs/ntfs/datarun/DataRunDecoder.java
Comment (tmyroadctfig) [inline src/main/java/org/jnode/fs/ntfs/datarun/DataRunDecoder.java:1]: Can you add in a test case for the changes here?

File: src/test/java/org/jnode/fs/ntfs/NTFSCompressedDataRunTriage4924Test.java [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+        byte[] compressed = FileSystemTestUtils.intArrayToByteArray(new int[]{
+                0x38, 0xb0, 0x88, 0x46, 0x23, 0x20, 0x00, 0x20,
+                0x47, 0x20, 0x41, 0x00, 0x10, 0xa2, 0x47, 0x01,
+                0xa0, 0x45, 0x20, 0x44, 0x00, 0x08, 0x45, 0x01,
+                0x50, 0x79, 0x00, 0xc0, 0x45, 0x20, 0x05, 0x24,
+                0x13, 0x88, 0x05, 0xb4, 0x02, 0x4a, 0x44, 0xef,
+                0x03, 0x58, 0x02, 0x8c, 0x09, 0x16, 0x01, 0x48,
+                0x45, 0x00, 0xbe, 0x00, 0x9e, 0x00, 0x04, 0x01,
+                0x18, 0x90, 0x00
+        });
```
Comment (tmyroadctfig) [inline src/test/java/org/jnode/fs/ntfs/NTFSCompressedDataRunTriage4924Test.java]: Just use the existing method here:
```
        byte[] compressed = FileSystemTestUtils.toByteArray(
                "38 b0 88 46 23 20 00 20" +
                "47 20 41 00 10 a2 47 01" +
                "a0 45 20 44 00 08 45 01" +
                "50 79 00 c0 45 20 05 24" +
                "13 88 05 b4 02 4a 44 ef" +
                "03 58 02 8c 09 16 01 48" +
                "45 00 be 00 9e 00 04 01" +
                "18 90 00");
```

File: src/main/java/org/jnode/fs/ntfs/datarun/CompressedDataRun.java [hunk 8]
```diff
@@ -162,105 +169,201 @@ public int readClusters(long vcn, byte[] dst, int dstOffset, int nrClusters, int
...
         // I tried using buffers but positions in those are always from the beginning and
         // I had to also maintain a position from the start of the current block.
         final OffsetByteArray compressedData = new OffsetByteArray(compressed);
         final OffsetByteArray uncompressedData = new OffsetByteArray(uncompressed);
 
-        for (int i = 0; i * BLOCK_SIZE < uncompressed.length; i++) {
-            final int consumed = uncompressBlock(compressedData, uncompressedData);
+        // "compressedData.offset < compressed.length - 1" means we can read in at least two bytes (the 16-bit header)
+        // TODO,
+        //  figure out why "compressedData.offset < compressed.length" is not the end condition.
```
Comment (david-sitsky) [inline src/main/java/org/jnode/fs/ntfs/datarun/CompressedDataRun.java]: It looks like this is your end condition now?  Should this comment be updated?

File: src/main/java/org/jnode/fs/ntfs/datarun/CompressedDataRun.java [hunk 9]
```diff
@@ -162,105 +169,201 @@ public int readClusters(long vcn, byte[] dst, int dstOffset, int nrClusters, int
...
+        //  It indicates that somewhere we didn't read the compressed array correctly before reaching here..
+        while (compressedData.offset < compressed.length - 1) {
+            // Bits [11:0] contain the size of the compressed chunk, minus three bytes.
```
Comment (david-sitsky) [inline src/main/java/org/jnode/fs/ntfs/datarun/CompressedDataRun.java:190]: minus three bytes?  You seem to be adding 3 bytes.  Also why is this required?  Is this code to try and round compressedChunkSize to be a 4 byte multiple or is it really 3 bytes required?

File: src/main/java/org/jnode/fs/ntfs/datarun/CompressedDataRun.java [hunk 10]
```diff
@@ -162,105 +169,201 @@ public int readClusters(long vcn, byte[] dst, int dstOffset, int nrClusters, int
...
+        //  Bits [14:12] contain a signature indicating the format of the subsequent data.
+        //  Bits [11:0] contain the size of the compressed chunk, minus three bytes.
         final int rawLen = compressed.getShort(cpos);
         cpos += 2;
+
+        // Bits 14 down to 12 contain a signature value. This value MUST always be 3 (unless the header denotes the end of the compressed buffer).
+        final int signature = rawLen & 0x7000;
+        if (rawLen != 0 && signature != 0x3000 && log.isDebugEnabled()) {
+            log.error("ntfs_uncompblock: signature {} is not 3", signature);
+        }
```
Comment (david-sitsky) [inline src/main/java/org/jnode/fs/ntfs/datarun/CompressedDataRun.java:240]: What is this code for.. should the log.isDebugEnabled() stay in the condition since you are using log.error?

File: src/main/java/org/jnode/fs/ntfs/datarun/CompressedDataRun.java [hunk 11]
```diff
@@ -162,105 +169,201 @@ public int readClusters(long vcn, byte[] dst, int dstOffset, int nrClusters, int
...
+                }
+
                 if ((ctag & 1) != 0) {
                     int j, lmask, dshift;
                     for (j = pos - 1, lmask = 0xFFF, dshift = 12;
-                         j >= 0x10; j >>= 1) {
+                         j >= 0x10; j >>>= 1) {
                         dshift--;
-                        lmask >>= 1;
+                        lmask >>>= 1;
```
Comment (david-sitsky) [inline src/main/java/org/jnode/fs/ntfs/datarun/CompressedDataRun.java:308]: Interesting this change to use >>> which makes sense.  It makes me wonder if there is any other code in thius module thta might be using >> instead of >>>.  Have you checked?

File: src/main/java/org/jnode/fs/ntfs/datarun/CompressedDataRun.java [hunk 12]
```diff
@@ -162,105 +169,201 @@ public int readClusters(long vcn, byte[] dst, int dstOffset, int nrClusters, int
...
-                    final int blen = Math.min(3 + (tmp & lmask), BLOCK_SIZE - pos);
 
-                    // Note that boff is negative.
-                    uncompressed.copyFrom(uncompressed, pos + boff, pos, blen);
-                    pos += blen;
+                    // While using the compressed buffers,
+                    // the stored displacement must be incremented by 1 and
+                    // the stored length must be incremented by 3,
+                    // to get the actual displacement and length.
+                    final int boff = (tmp >> dshift) + 1;
```
Comment (david-sitsky) [inline src/main/java/org/jnode/fs/ntfs/datarun/CompressedDataRun.java]: Should this be >>> ?

File: src/main/java/org/jnode/fs/ntfs/datarun/CompressedDataRun.java [hunk 13]
```diff
@@ -162,105 +169,201 @@ public int readClusters(long vcn, byte[] dst, int dstOffset, int nrClusters, int
...
+
+                    // TODO,
+                    //  no idea why the offset may be even larger than the (uncompressed.array.offset + pos),
+                    //  -- it apparently makes it no place to start to read the data.
+                    //  we didn't find any documentation explaining this case..
+                    //  just return as an error for now.
+                    if (boff > uncompressed.offset + pos) {
+                        log.error("Failed to decompress data, the offset (to start to back to read) {} exceeds the sum of " +
+                                "the current position {} and the uncompressed.offset {}", boff, pos, uncompressed.offset);
+                        return pos;
```
Comment (david-sitsky) [inline src/main/java/org/jnode/fs/ntfs/datarun/CompressedDataRun.java:335]: It seems like we should throw an exception here?  Logging and returning a valid RC will mean the upper-layers will continue to run as they think there is no error.

File: src/test/resources/org/jnode/fs/ntfs/libfsntfs/fsntfs_test_compression_uncompressed_data1.bin [hunk 14]
```diff
@@ -0,0 +1,166 @@
+		   GNU LESSER GENERAL PUBLIC LICENSE
+                       Version 3, 29 June 2007
```
Comment (david-sitsky) [inline src/test/resources/org/jnode/fs/ntfs/libfsntfs/fsntfs_test_compression_uncompressed_data1.bin:2]: I don't think you can put LGPLv3 code in there, since this model is Apache licensed, which is compatible.
